src/mito2/src/sst/parquet/reader.rs

```diff
@@ -513,166 +534,293 @@ impl RowGroupReaderBuilder {
-    async fn fetch_next_record_batch(&mut self) -> Result<Option<RecordBatch>> {
-        if let Some(row_group_reader) = &mut self.current_reader {
-            if let Some(record_batch) =
-                row_group_reader
-                    .next()
-                    .transpose()
-                    .context(ArrowReaderSnafu {
-                        path: self.reader_builder.file_path(),
-                    })?
-            {
-                return Ok(Some(record_batch));
-            }
+    #[cfg(test)]
+    pub fn parquet_metadata(&self) -> Arc<ParquetMetaData> {
+        self.context.reader_builder().parquet_meta.clone()
+    }
+}
+
+/// Reader to read a row group of a parquet file.
+pub(crate) struct RowGroupReader {
+    /// Context for file ranges.
+    context: FileRangeContextRef,
+    /// Inner parquet reader.
+    reader: ParquetRecordBatchReader,
+    /// Buffered batches to return.
+    batches: VecDeque<Batch>,
+    /// Local scan metrics.
+    metrics: Metrics,
+}
+
+impl RowGroupReader {
+    /// Creates a new reader.
+    pub(crate) fn new(context: FileRangeContextRef, reader: ParquetRecordBatchReader) -> Self {
+        Self {
+            context,
+            reader,
+            batches: VecDeque::new(),
+            metrics: Metrics::default(),
         }
+    }
 
-        // No more items in current row group, reads next row group.
-        while let Some((row_group_idx, row_selection)) = self.row_groups.pop_first() {
-            let mut row_group_reader = self
-                .reader_builder
-                .build(row_group_idx, row_selection)
-                .await?;
-            let Some(record_batch) =
-                row_group_reader
-                    .next()
-                    .transpose()
-                    .context(ArrowReaderSnafu {
-                        path: self.reader_builder.file_path(),
-                    })?
-            else {
-                continue;
+    /// Resets the parquet reader.
+    fn reset_reader(&mut self, reader: ParquetRecordBatchReader) {
+        self.reader = reader;
+    }
+
+    /// Tries to fetch next [Batch] from the reader.
+    async fn next_batch(&mut self) -> Result<Option<Batch>> {
+        if let Some(batch) = self.batches.pop_front() {
+            self.metrics.num_rows += batch.num_rows();
+            return Ok(Some(batch));
+        }
+
+        // We need to fetch next record batch and convert it to batches.
+        while self.batches.is_empty() {
+            let Some(record_batch) = self.fetch_next_record_batch()? else {
+                return Ok(None);
             };
+            self.metrics.num_record_batches += 1;
 
-            // Sets current reader to this reader.
-            self.current_reader = Some(row_group_reader);
-            return Ok(Some(record_batch));
+            self.context
+                .read_format()
+                .convert_record_batch(&record_batch, &mut self.batches)?;
```

Comment:
Can we concatenate the pruned batches into one batch, so that we can use an `Option` to hold the resulting batch and avoid yielding many fragmented batches?

Reply:
Concatenating batches might be slower than returning those batches directly.